math_utils.py

```diff
@@ -1,23 +1,39 @@
 from scipy.spatial.transform import Rotation
 
-def quaternion_to_pitch_and_roll(x, y, z, w):
-    quaternion = Rotation.from_quat([x, y, z, w])
-    eulerAngles = quaternion.as_euler('xzy', degrees=True)
-
-    pitch, roll, _ = eulerAngles
+last_pitch = 0
+safety_counter = 0
+safety_mode = False
 
-    # pitch = max(-15.5, min(pitch, 33.75))
-    # roll = max(-40, min(roll, 40))
+def quaternion_to_pitch_and_roll(x, y, z, w):
+    global last_pitch, safety_counter, safety_mode
 
+    quaternion = Rotation.from_quat([x, y, z, w])
+    euler_angles = quaternion.as_euler('xzy', degrees=True)
+
+    pitch, roll, _ = euler_angles
+    
+    if (pitch > 0 and pitch-last_pitch >= 0.5): # Will have to be changed with testing. "0.5" should be replaced with the large single-frame delta seen in the test where an interlock occurred.
+        safety_counter += 1
+    elif (pitch <= 0):
+        safety_mode = False
+        safety_counter = 0
+    
+    if (safety_counter >= 5): # Will also have to be changed in testing. Can modify the sensitivity of safety mode.
+        safety_mode = True
+        
+    if (safety_mode and pitch > 29.4): # Final thing that may need tweaking. 29.4 is the maximum angle achievable by the hardware, but orchestrated motion may make this more complicated.
```

Comment:
I refactored your nested `if` into this `and` statement for better readability, but it is effectively the same logic.

Question: Shouldn't this be an `or` instead of an `and`? According to my understanding, we want to decrease the pitch when:

1. the pitch changes very quickly, or when
2. the pitch reaches the hardware limit

I might be wrong though. Let me know what you think.

Reply:
My logic for this was once we enter safety mode, we have verified that a climb may be potentially unsafe. The interlock issue is that sometimes it whips up to a certain position so fast that it flings the hardware outside of the acceptable range. My idea here was that by dampening the angles in this climb, we will never be close to going out of the hardware range so can thus never be flung out of it.